### DISCUSSION TITLE
Js callbacks

### DISCUSSION THREAD

**Discussion by MishaKoval** [Voltstro-Studios/UnityWebBrowser#274] [Requests]
### What platform are you experiencing this issue on?

Windows

### What architecture is your platform?

64-Bit

### What version of UWB are you using?

Latest VoltstroUPM Release

### What Unity version are you running?

2021.3.26f1

### Describe what the issue you are experiencing is.

I try to read some data from html page in webView. I call ExecuteJs method and write something like that: 
var data = document.GetElementByClassName("example");
if(...)
{
    conlose.log("Have data");
}
And i use logger to reed messages from browser and check logs. But some js methods dont work like in usual browser.

### Provide reproducible steps for this issue.

private void Test()
        {
            webBrowserClient.ExecuteJs("var displayValue = document.getElementById('successPage');var data = displayValue.getElementsByClassName('payment_receipt__description');for (let index = 0; index < data.length; index++) {console.log('Result' +data[index].textContent);}");
        }
        

### Any additional info you like to provide?

public class JsExecutor : MonoBehaviour
    {
        [SerializeField] private BaseUwbClientManager clientManager;
        [SerializeField] private DataPrintToCheck dataPrintToCheck;
        private WebBrowserClient webBrowserClient;

        private bool isPrinted = false;
        
        private IEnumerator Start()
        {
            yield return null;
            yield return null;
            webBrowserClient = clientManager.browserClient;
            webBrowserClient.processLogHandler.OnProcessOutputLog += ProcessLogHandlerOnOnProcessOutputLog;
            StartCoroutine(CheckPageResult());
        }

        private void OnEnable()
        {
            //StartCoroutine(CheckPageResult());
        }

        private void OnDestroy()
        {
            webBrowserClient.processLogHandler.OnProcessOutputLog -= ProcessLogHandlerOnOnProcessOutputLog;
        }
        
        private readonly WaitForSeconds step = new(1);

        private IEnumerator CheckPageResult()
        {
            while (true)
            {
                if (!isPrinted)
                {
                    ExecuteJs();
                }
                yield return step;
            }
            // ReSharper disable once IteratorNeverReturns
        }

        [ContextMenu("TestSelectorJil")]
        private void TestJil()
        {
            webBrowserClient.ExecuteJs("var data = document.getElementsByClassName('menu-item menu-item-type-custom menu-item-object-custom menu-item-73 depth');console.log(data);");
        }
        

        [ContextMenu("TestSelector")]
        private void Test()
        {
            webBrowserClient.ExecuteJs("var displayValue = document.getElementById('successPage');var data = displayValue.getElementsByClassName('payment_receipt__description');for (let index = 0; index < data.length; index++) {console.log('Result' +data[index].textContent);}");
        }
        
        [ContextMenu("Refresh")]
        private void Refresh()
        {
            webBrowserClient.Refresh();
        }

        [ContextMenu("JS")]
        public void ExecuteJs()
        {
            Debug.Log("Execute");
            
            //webBrowserClient.Refresh();
            webBrowserClient.ExecuteJs("var displayValue = document.getElementById('successPage');console.log(displayValue.style.display);");
        }

        private List<string> resultData = new();
        
        private void ProcessLogHandlerOnOnProcessOutputLog(string obj)
        {
            JsonLogStructure logStructure = JsonConvert.DeserializeObject<JsonLogStructure>(obj);
            var message = logStructure.Message.Remove(0, 22);
            Debug.Log("Message: " + message);
            if (message == "none")
            {
                Debug.Log("NONE");
            }
            if (message.StartsWith("Result"))
            {
                resultData.Add(message.Remove(0,6));
                if (resultData.Count == 12)
                {
                    resultData.Add("0");
                    isPrinted = true;
                    dataPrintToCheck.SetupData(resultData);
                    dataPrintToCheck.Print();
                }
            }
            
            if (message == "block")
            {
                resultData.Clear();
                //webBrowserClient.ExecuteJs("var displayValue = document.getElementById('successPage');var data = displayValue.getElementsByClassName('payment_receipt__description');for (let index = 0; index < data.length; index++) {console.log('Result' + data[index].textContent);}");
                Debug.Log("Have");
            }
        }
        
        private class JsonLogSeverityConverter : JsonConverter
        {
            public override bool CanWrite => false;

            public override bool CanRead => true;

            public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
            {
                throw new NotImplementedException();
            }

            public override object ReadJson(JsonReader reader, Type objectType, object existingValue,
                JsonSerializer serializer)
            {
                if (reader.Value is string value)
                {
                    if (value == "Error")
                        return LogSeverity.Error;
                    if (value == "Warning")
                        return LogSeverity.Warn;
                    if (value == "Debug")
                        return LogSeverity.Debug;
                }

                return null;
            }

            public override bool CanConvert(Type objectType)
            {
                if (objectType == typeof(string))
                    return true;

                return false;
            }
        }
        
        [Serializable]
        public class JsonLogStructure
        {
            [JsonProperty("@t")] public DateTime Timestamp { get; set; }

            [JsonProperty("@m")] public string Message { get; set; }

            [JsonConverter(typeof(JsonLogSeverityConverter))]
            [JsonProperty("@l")]
            public LogSeverity Level { get; set; } = LogSeverity.Info;

            [JsonProperty("@x")] public string Exception { get; set; }

            [JsonProperty("@i")] public string EventId { get; set; }
        }
    }

**Comment by AP2020C** [comment]
I, too, am trying to get the source of the downloaded page.
And I haven't success.
Your code seems too long for such a simple function.
So I come to ask the developer to add:
-  a private delegate like: void OnLoadFinish(string url, string html)
- an event to inject js code into the page, e.g. string jsResult webBrowserClient.ExecuteJs(string js);
- an event that allows you to read the result of console.log (you can know when the js has finished working because if you have its result in the variable then it has finished)
